docs/userguide/mpi_apps.rst

```diff
@@ -3,33 +3,66 @@ MPI Apps
+Parsl simplifies this by composing the "launcher" command from the resources specified at the time
+each app is invoked.
+
+The broad strokes of a complete solution involves the following components:
+
+1. Configuring the :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` with:
+    ``enable_mpi_mode=True``
+2. Specify an MPI Launcher from one of the supported launchers ("aprun", "srun", "mpiexec") for the
+   :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` with: ``mpi_launcher="srun"``
+3. Specify the provider that matches your cluster, (eg. user ``SlurmProvider`` for Slurm clusters)
+4. Set the non-mpi launcher to :class:`~parsl.launchers.SimpleLauncher`
+5. Specify resources required by the application via ``resource_specification`` as shown below:
+
+
+.. code-block:: python
+
+    # Define HighThroughputExecutor(enable_mpi_mode=True, mpi_launcher="mpiexec", ...)
 
-HTEx and MPI Tasks
+    @bash_app
+    def lammps_mpi_application(infile: File, parsl_resource_specification: Dict):
+        # PARSL_MPI_PREFIX will resolve to `mpiexec -n 4 -ppn 2 -hosts NODE001,NODE002`
+        return f"$PARSL_MPI_PREFIX lmp_mpi -in {infile.filepath}"
+
+    # Resources in terms of nodes and how ranks are to be distributed are set on a per app
+    # basis via the resource_spec dictionary.
+    resource_spec = {
+        "NUM_NODES" = 2,
+        "RANKS_PER_NODE" = 2,
+    }
+    future = lammps_mpi_application(File('in.file'), resource_specification=resource_spec)
+
+
+HTEX and MPI Tasks
 ------------------
 
-The :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` (HTEx) is the default execution provider
-available through Parsl.
-Parsl Apps which invoke MPI code require a dedicated HTEx configured such that every Parsl app
-will have access to any of the nodes available within a block,
-and Apps that will invoke the MPI launcher with the correct settings.
+The :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` (HTEX) is the
+default executor available through Parsl.
+Parsl Apps which invoke MPI code require MPI specific configuration such that:
+
+1. All workers are started on the lead-node (mom-node in case of Crays)
+2. Resource requirements of Apps are propagated to workers who provision the required number of nodes from within the batch job.
+
 
 Configuring the Provider
 ++++++++++++++++++++++++
 
-Parsl must be configured to deploy workers on exactly one node per block.
-This part is simple.
-Instead of defining `a launcher <execution.html#launchers>`_ which will
-place an executor on each node in the block, simply use the :class:`~parsl.launchers.SimpleLauncher`.
+Parsl must be configured to deploy workers on exactly one node per block. This part is
+simple. Instead of defining a launcher which will place an executor on each node in the
+block, simply use the :class:`~parsl.launchers.SimpleLauncher`.
```

Comment:
Could we not ignore the launcher in MPI mode if it's not being used?

Reply:
I think this makes sense. I will bring this up in the Parsl user meeting.

Reply:
I've updated the docs to specify that only `SingleNodeLauncher` is allowed and have added asserts to the code to only allow that one launcher. `SingleNodeLauncher` is the default launcher so it is nice to support the default case with mpi_mode. If we have a new launcher in the future that could work with the mpi_mode, we can update that assert then.